service/history/workflow/mutable_state_impl.go

```diff
@@ -6252,3 +6259,259 @@ func (ms *MutableStateImpl) RefreshExpirationTimeoutTask(ctx context.Context) er
+	updatedUpdateInfos map[string]*persistencespb.UpdateInfo,
+	versionedTransition *persistencespb.VersionedTransition,
+) {
+	for updateID, ui := range updatedUpdateInfos {
+		if CompareVersionedTransition(ui.GetLastUpdateVersionedTransition(), versionedTransition) <= 0 {
+			continue
+		}
+		if existing, ok := ms.executionInfo.UpdateInfos[updateID]; ok {
+			ms.approximateSize -= existing.Size() + len(updateID)
+		} else {
+			ms.executionInfo.UpdateCount++
+		}
+		ms.executionInfo.UpdateInfos[updateID] = ui
+		ms.approximateSize += ui.Size() + len(updateID)
+		ms.updateInfoUpdated[updateID] = struct{}{}
+	}
+}
+
+func (ms *MutableStateImpl) syncExecutionInfo(current *persistencespb.WorkflowExecutionInfo, incoming *persistencespb.WorkflowExecutionInfo) {
+	ms.ClearStickyTaskQueue()
+
+	current.SubStateMachineTombstoneBatches = incoming.SubStateMachineTombstoneBatches
+	current.TransitionHistory = incoming.TransitionHistory
+
+	current.LastEventTaskId = incoming.LastEventTaskId
+	current.StateTransitionCount = incoming.StateTransitionCount
+
+	if ms.executionInfo.WorkflowTaskType == enumsspb.WORKFLOW_TASK_TYPE_SPECULATIVE {
+		ms.workflowTaskManager.deleteWorkflowTask()
+	} else {
+		taskInfo := &WorkflowTaskInfo{
+			Version:             incoming.WorkflowTaskVersion,
+			ScheduledEventID:    incoming.WorkflowTaskScheduledEventId,
+			StartedEventID:      incoming.WorkflowTaskStartedEventId,
+			RequestID:           incoming.WorkflowTaskRequestId,
+			WorkflowTaskTimeout: incoming.WorkflowTaskTimeout.AsDuration(),
+			Attempt:             incoming.WorkflowTaskAttempt,
+			StartedTime:         incoming.WorkflowTaskStartedTime.AsTime(),
+			ScheduledTime:       incoming.WorkflowTaskScheduledTime.AsTime(),
+
+			OriginalScheduledTime: incoming.WorkflowTaskOriginalScheduledTime.AsTime(),
+			Type:                  incoming.WorkflowTaskType,
+
+			SuggestContinueAsNew:   incoming.WorkflowTaskSuggestContinueAsNew,
+			HistorySizeBytes:       incoming.WorkflowTaskHistorySizeBytes,
+			BuildId:                incoming.WorkflowTaskBuildId,
+			BuildIdRedirectCounter: incoming.WorkflowTaskBuildIdRedirectCounter,
+		}
+		ms.workflowTaskManager.UpdateWorkflowTask(taskInfo)
+	}
+}
+
+func (ms *MutableStateImpl) syncSubStateMachinesByType(incoming map[string]*persistencespb.StateMachineMap) {
+	ms.executionInfo.SubStateMachinesByType = incoming
+	ms.mustInitHSM()
+}
+
+func (ms *MutableStateImpl) applyTombstones(tombstoneBatches []*persistencespb.StateMachineTombstoneBatch, versionedTransition *persistencespb.VersionedTransition) {
+	for _, tombstoneBatch := range tombstoneBatches {
+		if CompareVersionedTransition(tombstoneBatch.VersionedTransition, versionedTransition) <= 0 {
+			continue
+		}
+		for _, tombstone := range tombstoneBatch.StateMachineTombstones {
+			switch tombstone.StateMachineKey.(type) {
+			case *persistencespb.StateMachineTombstone_ActivityScheduledEventId:
+				ms.deleteActivityInfos[tombstone.GetActivityScheduledEventId()] = struct{}{}
```

Comment:
Check if the activity is still pending before adding to the delete map. Same for other sub state machines.